README.md
# humanify-numbers [![License: MIT](https://img.shields.io/badge/license-MIT-yellow.svg)](https://opensource.org/licenses/MIT) [![npm version](https://img.shields.io/npm/v/humanify-numbers.svg?style=flat)](https://www.npmjs.com/package/humanify-numbers) [![Run example](https://img.shields.io/static/v1?label=repl.it&message=run%20example&color=blue)](https://repl.it/join/ahkbmpzw-ezborgy)

Stringify numbers for humans, aka **_Humanify_** the numbers. 

E.g. 123456 is "123.4K" or "123.4 Thousands" or "120.5KiB"

## Install 

~~~
npm install -s humanify-numbers
~~~

## Use in TypeScript 

Simple TypeScript code:

~~~typescript
import { Numbers } from 'humanify-numbers'

console.log("1234567890 ➜", Numbers.stringify(1234567890))
// output: 1.2B

console.log("1234567890 ➜", Numbers.stringify(1234567890, 
                            { style: 'space' })) 
// output: 1.1GiB

console.log("1234567890 ➜", (1234567890).toLocaleString())
// compare with toLocalString(): 1,234,567,890
~~~

## Use in JavaScript

Same code in simple JavaScript:

~~~javascript
const { Numbers } = require("humanify-numbers");

console.log("1234 ➜", Numbers.stringify(1234)); 
// output: 1234 ➜ 1.2K
~~~

## Examples

### Number

|Value|Humanify|
|---|---|
|1234|1.2K|
|1234567|1.2M|
|1234567890|1.2B|
|1234567890123|1.2T|
|-1234|-1.2K|
|1.234|1.2|
|0.1234|0.1|
|abcd|NaN|

### space

Following the IEC 80000-13:2008 spec:

|Value|Humanify|
|---|---|
|1234|1.2KiB|
|1234567|1.1MiB|
|1234567890|1.1GiB|
|1234567890123|1.1TiB|
|1234567890123456|1PiB|
|1234567890123456789|1EiB|
|1.234|NaN|
|-1234|NaN|
|abcd|NaN|

### Percent

|Value|Humanify|
|---|---|
|0.01234|1.2%|
|0.001234|0.1%|
|0.1234|12.3%|
|1.234|123.4%|
|12.34|1234.0K%|

### namedNumber

|Value|Humanify|
|---|---|
|1000|1 Thousand|
|1234|1.2 Thousands|
|1000000|1 Million|
|1234567|1.2 Millions|

# Options

An optional parameter to manage how the numbers are humanified. 

|Option|Type|Description|
|---|---|---|
|style|string| "number" (default) or "space" or "percent" or "namedNumber"|
|precision|number| Value between 0 and 3, default 1|
|math|string| "floor" (default) or "round" or "ceil"|
|format (deprecated)|string| "number" (default) or "byte" or "percent"|

# Limits

* "precision": number between 0 and 3
* "number": Value < 1,000T (aka 10^15) and Value > -1,000T 
* "space": value must be a positive integer
* "space": Value < 1,000EiB (aka 1,024^7) and Value >= 0
* "percent": Value < 1,000K% (aka 10,000) and Value > -1,000K%


CHANGELOG.md
# Changelog

## 1.1.8 (Nov-28-2021)
* Fix more issues in readme.

## 1.1.7 (Nov-23-2021)
* Fix issues in readme.

## 1.1.6 (Nov-23-2021)
* Complete transition to XiB.
* Add option for named numbers, such that 1234 ➜ 1.2 Thousands

## 1.1.5 (May-15-2021)
* Change "storage" to "space",
* Change space strings from XB to XiB. E.g. 1234 used to be 1.2KB and will now be 1.2KiB (according to IEC 80000-13:2008).
* Add bigger space numbers up to 999EiB (< 10^21).

## 1.1.4 (April-24-2021)
* Change "format" to "style"
* Chance "byte" to "storage"
* For backward compatability, continue to support older terms. 

## 1.1.0 (March-24-2021)
* This is a breaking change!
* Suppress zeros after the decimal point if the number is already an integer.
* Change interface from Humanify to stringify.

## 1.0.1 (March-17-2021)
* Chnage folder structure. ./lib now contains the package code.

## 1.0.0 (March-6-2021)
* After more testing. Officially announce the release of 1.0.0.

## 0.0.19 (Feb-20-2021)
* shrink the dist folder by merging back all ts files. 

## 0.0.17 (Feb-19-2021)
* Add support math: floor, round, ceil. 
* Refactoring for better unification of percent and others.
* Add documentation of limites in the readme. 

## 0.0.16 (Feb-18-2021)
* Add support for percent format.
* Fix bugs when precision is not zero but there are no digits after the decimal point. 
* Fix bug with 1000.

## 0.0.13 (Feb-16-2021)
* Add replt.it badge.

## 0.0.12 (Feb-15-2021)
* Add tests to build.

## 0.0.10 (Feb-15-2021)
* Add testing framework. 
* Fix bug with 123.456 ➜ 123.5
* Change options format:"bytes" to "byte".
* Non integers in "byte" format should return "NaN".

## 0.0.9 (Feb-15-2021)
* Fixing bunch of small issues, especially when running Node JavaScript examples. 

## 0.0.3 (Feb-14-2021)
* First published version.

lib/src/index.ts
// //////////////////////////////////////////////
// humanify-numbers
// Convert numbers to human readbale / short strings.
// Developed by simplyCoders, 2021.

// ----------------------------------------------
// Supported styless.
const stylesAvailable = ['number', 'space', 'percent', 'byte', 'storage', 'namedNumber']
interface IStyles {
  [key: string]: any
}
const styles: IStyles = {
  number: {
    suffix: ['', 'K', 'M', 'B', 'T', 'Q'],
    min: 10 ** (-3),
    max: 10 ** 15,
  },
  namedNumber: {
    suffix: ['', ' Thousand', ' Million', ' Billion', ' Trillion', ' Quadrillion'],
    min: 10 ** (-3),
    max: 10 ** 15,
  },
  space: {
    suffix: ['B', 'KiB', 'MiB', 'GiB', 'TiB', 'PiB', 'EiB'],
    min: 10 ** (0),
    max: 10 ** 21,
  },
  percent: {
    suffix: ['%', 'K%'],
    min: 10 ** (-3),
    max: 10 ** 6,
  },
}

// ----------------------------------------------
// Supported math functions.
const mathsAvailable = ['floor', 'round', 'ceil']
interface IMaths {
  [key: string]: any
}
const maths: IMaths = {
  floor: Math.floor,
  round: Math.round,
  ceil: Math.ceil,
}

// ----------------------------------------------
// The exposrt: Numbers class
export class Numbers {
  static stringify(value: number, options?: any) {
    // defaults
    const defaultOpts = { style: 'number', precision: 1, math: 'floor' }

    // validate value
    if (typeof value !== 'number') {
      return 'NaN'
    }

    // analyze options
    const opts = defaultOpts
    if ((typeof options !== 'undefined')
      && (typeof options.style === 'string')
      && (stylesAvailable.includes(options.style))) {
      opts.style = options.style
    }

    if ((typeof options !== 'undefined') // for backward compatability
      && (typeof options.format === 'string')
      && (stylesAvailable.includes(options.format))) {
      opts.style = options.format
    }

    if (opts.style === 'byte') { // for backward compatability
      opts.style = 'space'
    }

    if (opts.style === 'storage') { // for backward compatability
      opts.style = 'space'
    }

    if ((typeof options !== 'undefined')
      && (typeof options.precision === 'number')
      && (options.precision >= 0)
      && (options.precision <= 3)) {
      opts.precision = options.precision
    }

    if ((typeof options !== 'undefined')
      && (typeof options.math === 'string')
      && (mathsAvailable.includes(options.math))) {
      opts.math = options.math
    }

    const val = (opts.style === 'percent') ? (value * 100) : value

    // end cases
    if (val === 0) {
      const zSuffix = styles[opts.style].suffix[0]
      const zeroVal = (0).toString()
      return zeroVal + zSuffix
    }

    const absVal = Math.abs(val)
    if ((absVal < styles[opts.style].min) || (absVal > styles[opts.style].max)) {
      return 'NaN'
    }

    if (opts.style === 'space' && (!Number.isInteger(val) || val < 0)) {
      return 'NaN'
    }

    // process the value
    const denominator = (opts.style === 'space') ? 1024 : 1000
    let logOfVal = Math.floor(Math.log(absVal + 0.00001) / Math.log(denominator))
    logOfVal = (logOfVal < 0) ? 0 : logOfVal // optimize

    let normalizedVal = val / (denominator ** logOfVal)
    if (Math.floor(normalizedVal)===0 && logOfVal > 0) { // optimize
      logOfVal -= 1
      normalizedVal = val / (denominator ** logOfVal)
    }

    const precisionedVal = maths[opts.math](normalizedVal * (10 ** opts.precision)) 
                         / (10 ** opts.precision)

    const humanifiedNumber = (precisionedVal).toString()

    const suffix = styles[opts.style].suffix[logOfVal]
    const plural = (opts.style==='namedNumber' && precisionedVal!==1 && suffix!=='') ? 's' : ''
    
    
    // build the final string
    return humanifiedNumber + suffix + plural
  }
}
